kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+                    ghost_state.borrow_mut().points_to_map.tracked_remove(1);
+                    ghost_state.borrow_mut().points_to_map.tracked_map_keys_in_place(
+                        Map::<nat, nat>::new(
+                            |j: nat| 1 <= j && j < ghost_state@.cells.len(),
+                            |j: nat| (j + 1) as nat,
+                        ),
+                    );
+                    ghost_state.borrow_mut().points_to_map.tracked_insert(0, head_points_to);
+                    assert forall|i: nat|
+                        1 <= i < ghost_state@.cells.len() implies well_formed_node(
+                        ghost_state,
+                        i,
+                    ) by {
+                        assert(old(ghost_state)@.cells[(i + 1) as int]
+                            == ghost_state@.cells[i as int]);
+                        assert(old(ghost_state)@.points_to_map.dom().contains((i + 1) as nat)
+                            == ghost_state@.points_to_map.dom().contains(i));
+                        assert(old(ghost_state)@.points_to_map[(i + 1) as nat]
+                            == ghost_state@.points_to_map[i]);
+                    }
+                }
+            },
+            None => {
+                proof {
+                    assert(ghost_state@.cells.len() == 1);
+                }
+
+                self.head.0.write(Tracked(&mut head_points_to), None);
+
+                proof {
+                    ghost_state.borrow_mut().points_to_map.tracked_insert(0, head_points_to);
+                }
+            },
+        }
+        remove
+    }
+
+    pub fn iter(&'a self, ghost_state: &Tracked<GhostState<'a, T>>) -> (res: ListIteratorV<'a, T>)
+        requires
+            ghost_state@.cells.len() >= 1,
+            forall|i: nat|
+                0 <= i < ghost_state@.cells.len()
+                    ==> #[trigger] ghost_state@.points_to_map.dom().contains(i)
+                    && ghost_state@.points_to_map[i].is_init() && ghost_state@.points_to_map[i].id()
+                    == ghost_state@.cells[i as int].id(),
+            forall|i: nat|
+                0 <= i < (ghost_state@.cells.len() - 1) as nat
+                    ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                    Option::Some(_) => true,
+                    Option::None => false,
+                },
+            match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+                Option::Some(_) => false,
+                Option::None => true,
+            },
+            ghost_state@.cells[0].id() == self.head.0.id(),
+        ensures
+            res.cur == ghost_state@.points_to_map[0].value(),
+            res.index@ == 0,
+    {
+        proof {
+            assert(ghost_state@.points_to_map.dom().contains(0));
+        }
+        ListIteratorV::<T>::new(self, ghost_state)
+    }
+}
+
+fn main() {
+}
+
```

Comment:
We should have an example of how to use the `List` similar to what they did [here](https://github.com/verus-lang/verus/blob/51b476c776372b5185f5d855053e1c3e9a00798c/source/rust_verify/example/doubly_linked.rs#L574) with assertions on the content of the list.

Reply:
Ok, but I think we need to implement the `ListNodeV` trait for a type. I currently have not thought it through about how to manage the `points_to` structs.

Reply:
okay let's try to merge this one first and then add an example as the next step

Reply:
Actually I had troubles creating an example that Verus would accept. 

If I do not use `Option` for `next` below, or I use `Box`, Verus will complain `datatype must have at least one non-recursive variant`. However, when I use Option, there are challenges in implementing `next` for the `ListNodeV` trait. There could not be a valid lifetime for `PLACEHOLDER`, and there cannot be panic or `unreachable!` because Verus does not support. In addition, I cannot think of a way to verify that `next_points_to` is indeed pointing to `next`, because there cannot be additional `requires` in traits implementation, and there is no way to access `next` in the `ListNodeV` declaration.

```rust
    struct I32Node<'a> {
        value: i32,
        next: Option<ListLinkV<'a, I32Node<'a>>>,
        next_points_to: Tracked<PointsTo<Option<&'a I32Node<'a>>>>,
    }

    impl<'a> I32Node<'a> {
        fn new(value: i32) -> I32Node<'a> {
            let (list_link, tracked_points_to) = ListLinkV::<I32Node<'a>>::empty();
            I32Node {
                value,
                next: Some(list_link),
                next_points_to: tracked_points_to,
            }
        }
    }

    exec static PLACEHOLDER: ListLinkV<'static, I32Node> = ListLinkV::<I32Node<'static>>::empty().0;

    impl<'a> ListNodeV<'a, I32Node<'a>> for I32Node<'a> {
        fn next(&'a self, next_points_to: Tracked<&PointsTo<Option<&'a I32Node<'a>>>>) -> (res: &'a ListLinkV<'a, I32Node<'a>>) {
            match &self.next {
                Some(next) => {
                    assume(next.0.id() == next_points_to@.id());
                    next
                },
                None => {
                    assume(false);
                    &PLACEHOLDER
                },
            }
        }
    }
```